CONTRIBUTING.md

```diff
@@ -17,3 +17,56 @@ The contributor roles for the new CDK construct library initiative are designed
+
+## Construct Design
+
+Construct design in this library has two primary goals:
+
+1. Maintain a high level of consistency across constructs to make them easy to use and understand.
+2. Ensure that little to no changes are required to merge constructs into the core AWS CDK library.
+
+To achieve these goals, the following design principles are followed:
+
+* Focus on only accepting L2 constructs to start. L3 constructs will come later as the library matures.
+* Follow the AWS CDK contributing guides
+* Use the AWS CDK library as a reference for design patterns
+
+Additionally, constructs [SHOULD](https://datatracker.ietf.org/doc/html/rfc2119#section-3):
+* Follow the standard CDK pattern of having all options on the third construct parameter (props).
+* Be named after the AWS resource they create.
+* Acronyms used in the construct name should not be entirely captialized. For example, `Vpc` instead of `VPC`.
+* The primary L1 (Cfn*) construct should have an 'id' of `Resource`.
+* Appropriate .grant*() methods should be implemented for IAM permissions.
+* If the resource must reside in a VPC, then the construct must require a VPC as a parameter.
+  * Additionally, the construct should have a `connections` property that allows for the addition of security groups and subnets.
+* All properties and methods should be properly documented with TypeScript docs.
+* Default values, overridden properties, and supplied resources should be properly test in Fine-Grained Assertion test.
+  * For example, if a resource requires an IAM Role, tests should exist that a default Role is created and that a Role can be provided and used appropriately.
+* Unit test coverage should be at least 100% for all constructs. Any inability to reach this goal should be clearly stated in the Pull Request.
+* Integration tests should be written, at least one for each construct but more is appreciated.
+  * Integration tests should be designed to be quick to execute. If this can't be achieved because it requires creating
+    a resource that takes a long time to create, then the integration test can be skipped.
+* L1 constructs should be created in protected methods to allow for easy extension of the construct. For example:
+```typescript
+export class MyConstruct extends Construct {
+  protected something: CfnSomething;
+  constructor(scope: Construct, id: string, private props: MyConstructProps) {
+    super(scope, id);
+
+    this.createResource();
+  }
+  
+  protected createResource() {
+    // Create the resource here
+    this.something = new CfnSomething(this, 'Resource', {
+      something: this.props.something,
+    });
+  }
+}
+```
```

Comment:
You're right, you don't see this in the CDK and this is one thing that has driven my crazy about it for quite a while. It's using OOP language but doesn't follow OOP principles and I would like to change that with this library. With the pattern I detailed here then it would be possible for others to subclass this construct and override just specific parts, still reusing all of the other functionality.

The storage of the instance variable is just so other constructs can easily reference it later. Whether it's the protected method that does the setting or the constructor that sets it doesn't really matter to me. Perhaps the later is better.

Reply:
I see. I respect your unique opinion of this library. Let's keep going.